File: src/utils/terminalCommandLine.ts
export interface ParsedTerminalCommandLine {
  executable: string | null;
  args: string[];
  error: string | null;
}

function repairLegacyWindowsQuotedCommandLine(commandLine: string): string {
  if (commandLine.startsWith('"')) {
    return commandLine;
  }

  const quoteCount = Array.from(commandLine).filter((char) => char === '"').length;
  if (quoteCount % 2 === 1) {
    const firstQuoteIndex = commandLine.indexOf('"');
    if (firstQuoteIndex > 2) {
      const prefix = commandLine.slice(0, firstQuoteIndex);
      const looksWindowsPath = /^[A-Za-z]:[\\/]/.test(prefix);
      const looksExecutable = /\.(exe|cmd|bat|com)$/i.test(prefix);
      if (looksWindowsPath && looksExecutable) {
        return `"${commandLine}`;
      }
    }
  }

  return commandLine;
}

function tokenizeCommandLine(input: string): { tokens: string[]; error: string | null } {
  const tokens: string[] = [];
  let current = '';
  let quote: '"' | "'" | null = null;

  for (let index = 0; index < input.length; index += 1) {
    const char = input[index];

    if (quote) {
      if (char === quote) {
        quote = null;
        continue;
      }

      if (char === '\\' && input[index + 1] === quote) {
        current += quote;
        index += 1;
        continue;
      }

      current += char;
      continue;
    }

    if (char === '"' || char === "'") {
      quote = char;
      continue;
    }

    if (/\s/.test(char)) {
      if (current.length > 0) {
        tokens.push(current);
        current = '';
      }
      continue;
    }

    current += char;
  }

  if (quote) {
    return { tokens: [], error: 'Terminal command contains an unmatched quote.' };
  }

  if (current.length > 0) {
    tokens.push(current);
  }

  return { tokens, error: null };
}

export function parseTerminalCommandLine(commandLine: string | null): ParsedTerminalCommandLine {
  const input = repairLegacyWindowsQuotedCommandLine(commandLine?.trim() ?? '');
  if (input.length === 0) {
    return { executable: null, args: [], error: null };
  }

  const windowsExecutableMatch = input.match(
    /^([A-Za-z]:[\\/].*?\.(?:exe|cmd|bat|com))(?=\s|$)(?:\s+(.*))?$/i,
  );
  if (windowsExecutableMatch) {
    const executable = windowsExecutableMatch[1] ?? null;
    const rest = windowsExecutableMatch[2]?.trim() ?? '';
    const tokenizedArgs = tokenizeCommandLine(rest);

    if (tokenizedArgs.error) {
      return { executable: null, args: [], error: tokenizedArgs.error };
    }

    return {
      executable,
      args: tokenizedArgs.tokens,
      error: null,
    };
  }

  const tokenized = tokenizeCommandLine(input);
  if (tokenized.error) {
    return { executable: null, args: [], error: tokenized.error };
  }

  const tokens = tokenized.tokens;

  if (tokens.length === 0) {
    return { executable: null, args: [], error: 'Terminal command is empty.' };
  }

  return {
    executable: tokens[0] ?? null,
    args: tokens.slice(1),
    error: null,
  };
}

File: src-tauri/src/core/terminal_command_line.rs
//! Terminal command-line parsing helpers.

/// Repairs a legacy Windows command line where a quoted path accidentally lost
/// its opening quote, for example `C:\Program Files\App\app.exe" --flag`.
pub fn repair_legacy_windows_quoted_command_line(command_line: &str) -> String {
    let input = command_line.trim();
    if input.is_empty() || input.starts_with('"') {
        return input.to_string();
    }

    let quote_count = input.chars().filter(|ch| *ch == '"').count();
    if quote_count % 2 == 1 {
        if let Some(first_quote_index) = input.find('"') {
            let prefix = &input[..first_quote_index];
            let looks_windows_path = prefix.len() >= 3
                && prefix.as_bytes()[1] == b':'
                && matches!(prefix.as_bytes()[2], b'\\' | b'/');
            let lower_prefix = prefix.to_ascii_lowercase();
            let looks_executable = [".exe", ".cmd", ".bat", ".com"]
                .iter()
                .any(|extension| lower_prefix.ends_with(extension));

            if looks_windows_path && looks_executable {
                return format!("\"{input}");
            }
        }
    }

    input.to_string()
}

fn split_windows_executable_with_spaces(command_line: &str) -> Option<(&str, &str)> {
    let input = command_line.trim();
    let bytes = input.as_bytes();
    if bytes.len() < 4
        || !bytes[0].is_ascii_alphabetic()
        || bytes[1] != b':'
        || !matches!(bytes[2], b'\\' | b'/')
    {
        return None;
    }

    let lower = input.to_ascii_lowercase();
    for (index, _) in lower.match_indices('.') {
        for extension in [".exe", ".cmd", ".bat", ".com"] {
            if !lower[index..].starts_with(extension) {
                continue;
            }
            let executable_end = index + extension.len();
            let rest = &input[executable_end..];
            if rest.is_empty() || rest.chars().next().is_some_and(char::is_whitespace) {
                return Some((&input[..executable_end], rest.trim()));
            }
        }
    }

    None
}

fn tokenize_command_line(command_line: &str) -> Result<Vec<String>, String> {
    let repaired = repair_legacy_windows_quoted_command_line(command_line);
    let input = repaired.trim();
    if input.is_empty() {
        return Err("Terminal profile command line is empty".to_string());
    }

    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut quote: Option<char> = None;
    let mut chars = input.chars().peekable();

    while let Some(ch) = chars.next() {
        if let Some(active_quote) = quote {
            if ch == active_quote {
                quote = None;
                continue;
            }

            if ch == '\\' && chars.peek() == Some(&active_quote) {
                current.push(active_quote);
                chars.next();
                continue;
            }

            current.push(ch);
            continue;
        }

        if ch == '"' || ch == '\'' {
            quote = Some(ch);
            continue;
        }

        if ch.is_whitespace() {
            if !current.is_empty() {
                tokens.push(std::mem::take(&mut current));
            }
            continue;
        }

        current.push(ch);
    }

    if quote.is_some() {
        return Err("Terminal profile command line contains an unmatched quote".to_string());
    }

    if !current.is_empty() {
        tokens.push(current);
    }

    if tokens.is_empty() {
        return Err("Terminal profile command line is empty".to_string());
    }

    Ok(tokens)
}

pub fn parse_terminal_command_line(command_line: &str) -> Result<(String, Vec<String>), String> {
    let repaired = repair_legacy_windows_quoted_command_line(command_line);
    let input = repaired.trim();
    if let Some((executable, rest)) = split_windows_executable_with_spaces(input) {
        let args = if rest.is_empty() {
            Vec::new()
        } else {
            tokenize_command_line(rest)?
        };
        return Ok((executable.to_string(), args));
    }

    let tokens = tokenize_command_line(input)?;
    let executable = tokens
        .first()
        .cloned()
        .ok_or_else(|| "Terminal profile command line is empty".to_string())?;
    Ok((executable, tokens.into_iter().skip(1).collect()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_quoted_windows_path() {
        let (executable, args) =
            parse_terminal_command_line(r#""C:\Program Files\Git\bin\bash.exe" --login -i"#)
                .unwrap();

        assert_eq!(executable, r#"C:\Program Files\Git\bin\bash.exe"#);
        assert_eq!(args, vec!["--login".to_string(), "-i".to_string()]);
    }

    #[test]
    fn parses_unquoted_legacy_windows_path_with_spaces() {
        let (executable, args) =
            parse_terminal_command_line(r#"C:\Program Files\Git\bin\bash.exe --login -i"#).unwrap();

        assert_eq!(executable, r#"C:\Program Files\Git\bin\bash.exe"#);
        assert_eq!(args, vec!["--login".to_string(), "-i".to_string()]);
    }

    #[test]
    fn repairs_missing_opening_quote_for_legacy_windows_path() {
        let (executable, args) =
            parse_terminal_command_line(r#"C:\Program Files\Git\bin\bash.exe" --login -i"#)
                .unwrap();

        assert_eq!(executable, r#"C:\Program Files\Git\bin\bash.exe"#);
        assert_eq!(args, vec!["--login".to_string(), "-i".to_string()]);
    }

    #[test]
    fn rejects_unmatched_quote() {
        let result = parse_terminal_command_line(r#""/bin/zsh"#);

        assert!(result.is_err());
    }
}
